SOLR-14660 - move HDFS to a module

warperwolf opened the pull request:
https://issues.apache.org/jira/browse/SOLR-14660

# Description

This PR moves HDFS from Core to a Contrib module.

Note: I kept separate commits to make it easier to review and to change / revert if necessary. They can be squashed once everything is polished and we have a green light.

# Solution
- A new contrib module has been created with its own build
- HDFS classes have been moved to the new module
- This PR does not yet take advantage of the Solr packaging system, this is just a separate module. Implementing as a package will be a next step. 
- HDFS directory direct references have been removed from Solr core (for example the UpdateHandler had a direct reference to HdfsDirectoryFactory, removed it so it works the same way as other directory factories and can also be loaded as a plugin (core.getResourceLoader().newInstance(ulogPluginInfo, UpdateLog.class, true);)
- Solr Core still has some Hadoop-related (but non HDFS-related) classes (mostly Hadoop authentication), these are still left in Core , if they need to be separated, they will be handled in a separate jira
- The block cache feature was only used by HDFS, moved to the contrib module)
- The only change needed to use this module is to include its jar on the classpath (for example symlinking the jar to the web-inf/lib of the webapp), no changes to the solrconfig.xml are needed.
- DirectoryFactory had DirectoryFactory.java: public final static String LOCK_TYPE_HDFS   = "hdfs";  , but this was only referenced from the tests and from HdfsDirectoryFactory, removed it.
- Removed the deprecated flags from the code and the warning from the ref guide.
- Changed reference in ref guide to the new location of HdfsDirectoryFactory.
- Added a simple readme markdown doc 
 
# Tests
- Existing tests have been moved to the new contrib module where needed
- Test have been refactored where a Hdfs test was extending a Core test class. For example, HdfsChaosMonkeyNothingIsSafeTest (hdfs test) extends ChaosMonkeyNothingIsSafeTest (core test). 
Gradle does not allow dependencies between test projects of 2 different modules, so in such cases I introduced an abstract base class (here AbstractChaosMonkeyNothingIsSafeTest) which is extended by both. A good place to store these classes would be the test fixtures feature of gradle which is especially designed for this purpose, however some of the gradle plugins Solr uses fail because they are not compatible with fixtures (see https://issues.apache.org/jira/browse/SOLR-14660?focusedCommentId=17409690&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17409690 ), so I moved these classes to the Solr Test Framework project. 
- Solr core had some helper classes which are core related but they are also referenced by the hdfs classes (for example the MockCoreContainer.java) - moved those to gradle test fixtures of the Core project, and later moved them to the test framework
- I also created a Cloudera CDP test release and built a cluster with HDFS and Solr and tested there
- TestBackupRepositoryFactory had some references to HdfsBackupRepository, but the test case is not hdfs related. Removed that part.
- HDFSRecoveryZKTest was also failing before this change. Added an AwaitsFix annotation, needs to be fixed separately.


# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

madrob commented on solr/contrib/README.md:
```diff
@@ -0,0 +1,31 @@
+Apache Solr HDFS Contrib module
```
Should this be under the hfs contrib directory instead of top level?

risdenk replied:
Agree w/ @madrob - there is already a `solr/contrib/hdfs/README.md` as well? Should be consolidated.

warperwolf replied:
Fixed (https://github.com/apache/solr/pull/324/commits/e0a4b5bcd355554e2a59fc240123338e05a97716)

madrob commented on gradle/documentation/render-javadoc.gradle:
```diff
@@ -151,8 +151,25 @@ configure(project(":solr:test-framework")) {
         "org.apache.solr.analysis",
         "org.apache.solr.cloud",
         "org.apache.solr.core",
+        "org.apache.solr.handler",
```
Not sure I understand what changes caused this to start failing?

risdenk replied:
Agree with @madrob here

risdenk replied:
Ok so I looked into this. A handful of classes were moved from solr core into solr test-framework due to limitations in gradle not being able to share test classes. The move ended up in a few more split packages - which causes issues in the render-javadoc. So these exclusions are due to the new split packages. I didn't see a simple way to address this related to this change. 

I think test-framework should probably look at how it does packages (maybe need to at `.test.` or something to the package name) to avoid split packages.

madrob commented on solr/test-framework/src/java/org/apache/solr/handler/package.html:
```diff
@@ -0,0 +1,24 @@
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-->
+<!-- not a package-info.java, because we already defined this package in core/ -->
```
We probably want to look at renaming some of the packages for the moved classes, since split packages are going to cause problems down the line anyway. Maybe ok to do this in a follow on issue if it ends up being too much here.

risdenk replied:
I think this should be handled as a follow up. There are a bunch of existing split packages in `test-framework` which sadly this PR adds to, but follows the existing convention.

madrob reviewed (commented):
79/447 files viewed (not in order). Left some inline comments that caught my eye.

it is surprising to me that we generally needed to copy so much from test-files/collection1/conf to make the tests work, is there a chance that we don't need all of that?

dsmiley commented:
> The block cache feature was only used by HDFS, moved to the contrib module)

Is it _possible_ to use it in some other/generic Directory impl?  I haven't looked at it closely but in my mind it was fundamentally intertwined with HDFS but maybe I'm wrong.

risdenk commented on solr/core/src/test/org/apache/solr/core/TestBackupRepositoryFactory.java:
```diff
@@ -132,13 +131,7 @@ public void testRepositoryConfig() {
-      attrs.put(FieldType.CLASS_NAME, HdfsBackupRepository.class.getName());
-      attrs.put("location", "/tmp");
-      plugins[1] = new PluginInfo("repository", attrs);
-    }
```
Not sure I understand this removal? This changes the test for no reason?

risdenk commented on gradle/validation/rat-sources.gradle:
```diff
@@ -102,6 +102,12 @@ allprojects {
                     exclude "src/test-files/META-INF/services/*"
                     break
 
+                case ":solr:contrib:hdfs":
```
Agreed this should be cleaned up. I'd be surprised if any of these test files are needed for HDFS.

risdenk commented on gradle/documentation/render-javadoc.gradle:
```diff
@@ -151,8 +151,25 @@ configure(project(":solr:test-framework")) {
+        "org.apache.solr.handler",
         "org.apache.solr.handler.component",
         "org.apache.solr.update.processor",
+        "org.apache.solr.util",
+        "org.apache.solr.search.similarities",
+        "org.apache.solr.search.function",
+        "org.apache.solr.search.facet",
+        "org.apache.solr.schema"
```
Is this needed?

risdenk replied:
Ok so I looked into this. A handful of classes were moved from solr core into solr test-framework due to limitations in gradle not being able to share test classes. The move ended up in a few more split packages - which causes issues in the render-javadoc. So these exclusions are due to the new split packages. I didn't see a simple way to address this related to this change. 

I think test-framework should probably look at how it does packages (maybe need to at `.test.` or something to the package name) to avoid split packages.

risdenk commented on solr/contrib/hdfs/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# IDEA project file
+hdfs.iml
```
Agreed this should be removed.

warperwolf replied:
Done (https://github.com/apache/solr/pull/324/commits/e0a4b5bcd355554e2a59fc240123338e05a97716)

risdenk commented on solr/contrib/hdfs/build.gradle:
```diff
@@ -0,0 +1,56 @@
+  testImplementation project(':solr:test-framework')
+
+  // hadoop dependencies for tests
+  testImplementation ('org.apache.hadoop:hadoop-hdfs') { transitive = false }
```
Agreed we should remove the `transitive = false` since it causes some pain.

risdenk replied:
I've been looking at this as part of PR #553 and some other related dependency improvements. It doesn't make sense to "fix" this as part of this move. This should be done post moving to a module.

risdenk reviewed (changes requested):
I left some specific comments.

I don't understand the need for the Abstract***Base` changes. I understand the HDFS test ends up extending the tests, I don't know if its necessary to make them abstract or change the file names. Git seems to be confused between the remove/add and doesn't label these changes as a move. This would SIGNIFICANTLY reduce the diff here.

risdenk commented:
BTW I went through all the files and there aren't THAT many files that need to be reviewed for real changes.
* ~344/447 are just test-files - a lot could be removed?
* ~57/447 - are primarily just only moves
* 46 are real changes

dsmiley commented:
Note that the build.gradle of solr-core has since been improved by me to more clearly organize the dependencies and to use transitive=false much less.

madrob commented:
I'm concerned about this missing changes to the Hadoop code from the endianness fix, the upgrade to 3.3.1, and whatever else we've had to patch up since this PR was originally opened. What is our strategy for keeping things in sync.

risdenk commented:
> I'm concerned about this missing changes to the Hadoop code from the endianness fix, the upgrade to 3.3.1, and whatever else we've had to patch up since this PR was originally opened. What is our strategy for keeping things in sync.

@warperwolf and I have been recently talking about this. I've been keeping an eye on the changes that have been made to the Hadoop code and this PR so it shouldn't be terrible to keep up to date. We should get this merged as soon as reasonably ready to avoid having to keep it in sync any longer.

warperwolf commented on solr/core/src/java/org/apache/solr/update/UpdateHandler.java:
```diff
@@ -115,12 +115,9 @@ public UpdateHandler(SolrCore core, UpdateLog updateLog)  {
-                core.getResourceLoader().newInstance(ulogPluginInfo, UpdateLog.class, true);
-      }
+
+      ulog = ulogPluginInfo.className == null ? new UpdateLog():
```
I had a discussion with @risdenk on this one. Indeed the code currently picks the standard UpdateLog.class unless the sorlconfig.xml explicitely specifies the class of the updateLog definition. This can lead to the update logs being written to the local FS while the index of the same collection is being written to HDFS. While this is technically possible (it worked during my basic tests), this is likely not what we want. 
The Solr tests use the solr.tests.ulog variable for this, I took care of the correct setting for Hdfs tests.
<updateLog class="${solr.tests.ulog:solr.UpdateLog}">
This should be documented. 
With Kevin we also discussed that we could look into the Hdfs directory factory intialization and verify if we can check if an inconsistent update log class is used and error out in this case. I checked the directory factory code and to me it seems that we have no access to the update log type there, likely because they can also be defined per update handler and initialized later.
I think we could also consider the following: add a new method like getDefaultUpdateLogClassName to the DirectoryFactory class. This method which would return "solr.UpdateLog" in the base class, and the HdfsDirectoryFactory could override this by returning "solr.HdfsUpdateLog". In the UpdateHandler part, if we detect that the solrconfig did not specify any class, we could fall back to the one returned by this method. This would ensure compatibility with the existing solrconfigs for HDFS and non-HDFS configs (less pain during an upgrade), and at the same time it would allow the user to override the value in SolrConfig. @risdenk , @dsmiley what do you think?

dsmiley replied:
We discussed this in Slack; either way is fine.

warperwolf replied:
Done in https://github.com/apache/solr/pull/324/commits/9ae46ef4588359f1f8d468120c2cc5b5d20f335b

risdenk commented on solr/test-framework/src/java/org/apache/solr/cloud/api/collections/AbstractCollectionsAPIDistributedZkTestBase.java:
```diff
@@ -0,0 +1,645 @@
+ */
+@Slow
+public abstract class AbstractCollectionsAPIDistributedZkTestBase extends SolrCloudTestCase {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
```
@sonatype-lift ignore

risdenk commented on solr/test-framework/src/java/org/apache/solr/cloud/hadoop/HadoopTestUtil.java:
```diff
@@ -0,0 +1,126 @@
+import org.slf4j.LoggerFactory;
+
+public class HadoopTestUtil {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
```
@sonatype-lift ignore

risdenk commented on solr/contrib/hdfs/src/java/org/apache/solr/core/backup/repository/HdfsBackupRepository.java:
```diff
@@ -46,10 +46,6 @@
- */
-@Deprecated
 public class HdfsBackupRepository implements BackupRepository {
   private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
```
@sonatype-lift ignore

risdenk commented on solr/contrib/hdfs/src/java/org/apache/solr/store/hdfs/HdfsFileWriter.java:
```diff
@@ -30,9 +30,7 @@
 
 /**
  * @lucene.experimental
```
@sonatype-lift ignore

risdenk commented on solr/contrib/hdfs/src/java/org/apache/solr/update/HdfsUpdateLog.java:
```diff
@@ -44,9 +44,7 @@
 
 /**
  * @lucene.experimental
```
@sonatype-lift ignore

risdenk commented on solr/contrib/hdfs/src/java/org/apache/solr/store/hdfs/HdfsFileWriter.java:
```diff
@@ -30,9 +30,7 @@
 
 /**
  * @lucene.experimental
```
@sonatype-lift ignore

risdenk commented on solr/contrib/hdfs/src/java/org/apache/solr/update/HdfsUpdateLog.java:
```diff
@@ -44,9 +44,7 @@
 
 /**
  * @lucene.experimental
```
@sonatype-lift ignore

warperwolf commented:
> > (me) I'm most concerned with the apparent duplication of tons of Solr test configs. How can we avoid that?
> 
> I'd be a hack but I wonder if this contrib could declare a test resource "input" pointing to a sub-directory of solr-core test configs? I'd take almost any hack over duplicating all that test config.

@dsmiley @risdenk @madrob in commit https://github.com/apache/solr/pull/324/commits/2ec7d59195238473a7e6e6e7c5fee0112cf20f40 I deleted all test resource files from the hdfs module and I added a gradle step to it which copies all the resource files from solr core's test-files to the build/resources/test directory of the hdfs module. A better solution could be to modify the test classes of the HDFS module to point to the resource files in the core's test-files directory, however the test base classes like SolrTestCaseJ4 load resources through the class loader (so that they end up being loaded from the build/resources/test directory of the module we are testing at the moment), and this resource loading seems to occur at various locations, so it would take some time to tweak these.  For now I'd suggest going with this resource copying solution and change the tests later in a separate ticket. (File copying occurs only on hdfs test execution, and this resolves the original issue of having to maintain 2 set of resource files in git.)

risdenk commented on solr/test-framework/build.gradle:
```diff
@@ -31,5 +31,16 @@ dependencies {
   implementation 'io.dropwizard.metrics:metrics-jetty9'
   implementation 'com.lmax:disruptor'
+
+  // used by the hadoop-specific test framework classes
```
This has been cleaned up.

risdenk commented:
So current status as I see it:
* render-javadoc needs a little more attention - basically shouldn't need so many changes
* `Abstract*Base` test files moved to test-framework - not sure they are needed, but its a lot of the changes in this besides the move.

I think all of the other comments from @dsmiley and @madrob have been addressed by @warperwolf 

I'm planning to look into the above a bit more tomorrow. @warperwolf, @dsmiley, @madrob any other comments/thoughts/concerns? Any other blockers to getting this merged?

dsmiley commented on solr/core/src/java/org/apache/solr/update/UpdateHandler.java:
```diff
@@ -115,12 +115,12 @@ public UpdateHandler(SolrCore core, UpdateLog updateLog)  {
+      // if the update log class is not defined in the plugin info / solrconfig.xml
+      // (like <updateLog class="${solr.ulog:solr.UpdateLog}"> )
+      // we fall back use the one which is the default for the given directory factory
+      ulog = ulogPluginInfo.className == null ? dirFactory.newDefaultUpdateLog() :
```
but then uLogPluginInfo would be ignored, which might have settings to be applied.  Either `newDefaultUpdateLog` should take the pluginInfo as a parameter, or just have it return the class (or class name) to be instantiated (and maybe tweak the name to be getDefaultUpdateLogClassName).  I think the latter is cleaner.

risdenk replied:
@dsmiley  not sure I understand your comment here. The original logic:

```
DirectoryFactory dirFactory = core.getDirectoryFactory();
      if (dirFactory instanceof HdfsDirectoryFactory) {
        ulog = new HdfsUpdateLog(((HdfsDirectoryFactory)dirFactory).getConfDir());
      } else {
        ulog = ulogPluginInfo.className == null ? new UpdateLog():
                core.getResourceLoader().newInstance(ulogPluginInfo, UpdateLog.class, true);
      }
```

If `ulogPluginInfo.className == null` then `new UpdateLog();`. So I don't think there is any change here to the logic for `UpdateLog()`. Line 132 has `ulog.init(ulogPluginInfo);` which I think makes sure the ulog implementation takes into account the `uLogPluginInfo).

dsmiley replied:
Ooh; I missed line 132.  Thanks.

dsmiley commented on solr/modules/hdfs/README.md:
```diff
@@ -0,0 +1,31 @@
+
+Usage
+-----
+To use the module, it needs to be placed to the Solr web application classpath (for example by symlinking it to the WEB-INF/lib directory.)
```
IMO, unless there are unique constraints about HDFS's JARs, we can simply refer to the ref guide (`solr-plugins.adoc`)  on how plugins are installed.  There are a number of options.

risdenk replied:
Addressed in bb216a74511e8a13787556ff94da35332ca71692

dsmiley commented on solr/modules/hdfs/src/test/org/apache/solr/store/blockcache/BlockCacheTest.java:
```diff
@@ -257,7 +259,7 @@ public void testCacheConcurrent() throws Exception {
         }
       }
-      assertEquals("cache key differs from value's key", k, (Long) v.key);
+      Assert.assertEquals("cache key differs from value's key", k, (Long) v.key);
```
minor: you could just add static imports for this stuff.  I'm not sure why these changes are happening.

risdenk replied:
This can be cleaned up. Will push a fix shortly. Looks like its only the `test/org/apache/solr/store` tests

risdenk replied:
Fixed in e75d6ac1bc6c6ee7777f23ea0b2b8877928b86e6

dsmiley commented on solr/solr-ref-guide/src/solr-on-hdfs.adoc:
```diff
@@ -16,11 +16,12 @@
 
 Solr has support for writing and reading its index and transaction log files to the HDFS distributed filesystem.
 
+Starting from 9.0, this functionality has been moved from Solr Core to the HDFS contrib module. 
```
Not "contrib"

risdenk replied:
Fixed in 9675773a7e4ebdd2a794a021db187576571d2db8

dsmiley reviewed (commented):
It's great to see this so close now!

risdenk commented on gradle/validation/validate-source-patterns.gradle:
```diff
@@ -108,6 +108,11 @@ allprojects {
+}
+
+configure(project(':solr:modules:hdfs')) {
+  project.tasks.withType(ValidateSourcePatternsTask) {
```
Actually this can be fixed. There were previously `//nowarn` but needs to be `// nowarn_valid_logger` apparently now. Will push the fix shortly.

risdenk replied:
Fixed in c34e6a9376c4b3bc681ffd45a7829cf710ba82e5

risdenk commented on solr/contrib/README.md:
```diff
@@ -0,0 +1,31 @@
+
+Usage
+-----
+To use the module, it needs to be placed to the Solr web application classpath (for example by symlinking it to the WEB-INF/lib directory.)
```
Yea this needs to be addressed. I'm marking this as resolved since you have another comment on the correct file location. So its not going to get lost.

risdenk commented:
I went through these two and I'm not sure there is a good way to address them right now. I don't think it should hold up the merge of this PR.

* render-javadoc needs a little more attention - basically shouldn't need so many changes
    * This is related to the bullet below. There were a few classes moved to `test-framework` which ended up in a few more split packages. This should be addressed for all of test-framework.
* Abstract*Base test files moved to test-framework - not sure they are needed, but its a lot of the changes in this besides the move.
    * This seems to be necessary due to how gradle and test dependencies end up interacting. It matches the existing pattern of some test-framework classes. The commit log has details about why each file was created/moved in test-framework.

madrob commented:
I ran `gradlew assemble` and took a look at the output in `solr/packaging/build`... There were only 3 jars in `modules/hdfs/lib/` - I'm not sure what I expected but for some reason I imagined there would be more?

Hadoop-auth, -annotations, and -common are all still in WEB-INF/lib, do we need all of those? I guess auth depends on common and we haven't moved that out yet?

madrob commented:
I've been looking through and I _think_ that we can exclude Hadoop-annotations and leave it in only as a test dependency? I'm not sure.

warperwolf commented on solr/modules/hdfs/build.gradle:
```diff
@@ -0,0 +1,93 @@
+// schemas and SolrConfigs. This can be improved later by making the test classes
+// load the resources from core directories directly.
+task copySolrCoreTestResources(type: Copy) {
+  from('../../core/src/test-files') {
```
Done in https://github.com/apache/solr/pull/324/commits/843fc717695608b2f41dc2c8e4f8417e5a183344

risdenk commented:
@madrob I'm with you I expected more jars to be there in `hdfs/lib`, but `hadoop-auth`, `hadoop-annotations`, and `hadoop-common` are all needed by the existing Hadoop authentication logic. 

FWIW removing `hadoop-annotations` fails with:

```
> Task :solr:core:compileJava
/Users/risdenk/.gradle/caches/modules-2/files-2.1/org.apache.hadoop/hadoop-common/3.2.0/e47a88c42c450e6e4b23bf951356c203cae2db24/hadoop-common-3.2.0.jar(/org/apache/hadoop/fs/FileSystem.class): warning: Cannot find annotation method 'value()' in type 'LimitedPrivate': class file for org.apache.hadoop.classification.InterfaceAudience not found
/Users/risdenk/.gradle/caches/modules-2/files-2.1/org.apache.hadoop/hadoop-common/3.2.0/e47a88c42c450e6e4b23bf951356c203cae2db24/hadoop-common-3.2.0.jar(/org/apache/hadoop/fs/FileSystem.class): warning: Cannot find annotation method 'value()' in type 'LimitedPrivate'
/Users/risdenk/.gradle/caches/modules-2/files-2.1/org.apache.hadoop/hadoop-common/3.2.0/e47a88c42c450e6e4b23bf951356c203cae2db24/hadoop-common-3.2.0.jar(/org/apache/hadoop/fs/Path.class): warning: Cannot find annotation method 'value()' in type 'LimitedPrivate'
/Users/risdenk/.gradle/caches/modules-2/files-2.1/org.apache.hadoop/hadoop-common/3.2.0/e47a88c42c450e6e4b23bf951356c203cae2db24/hadoop-common-3.2.0.jar(/org/apache/hadoop/security/UserGroupInformation.class): warning: Cannot find annotation method 'value()' in type 'LimitedPrivate'
/Users/risdenk/.gradle/caches/modules-2/files-2.1/org.apache.hadoop/hadoop-common/3.2.0/e47a88c42c450e6e4b23bf951356c203cae2db24/hadoop-common-3.2.0.jar(/org/apache/hadoop/security/UserGroupInformation.class): warning: Cannot find annotation method 'value()' in type 'LimitedPrivate'
/Users/risdenk/.gradle/caches/modules-2/files-2.1/org.apache.hadoop/hadoop-auth/3.2.0/b1b95aed9aa956ffb7d21e30a0415ca14d91c4ad/hadoop-auth-3.2.0.jar(/org/apache/hadoop/security/authentication/util/KerberosName.class): warning: Cannot find annotation method 'value()' in type 'LimitedPrivate'
error: warnings found and -Werror specified
```

risdenk commented on solr/core/build.gradle:
```diff
@@ -146,11 +146,6 @@ dependencies {
   testImplementation 'org.apache.commons:commons-compress'
   testImplementation ('com.sun.jersey:jersey-servlet') { transitive = false }
 
-  // HDFS
```
So sadly no we can't straight up remove them. I did however reorganize them under `hadoop-auth` to be clear and added to the HDFS module the necessary runtimeOnly dependencies for when `hadoop-auth` does get removed from core - we don't break the HDFS module.

Addressed in ae67b144357113a9ee80ab981ee1279242c53242

risdenk commented:
Thanks @madrob and @dsmiley. I think your comments have been addressed by @warperwolf and I. Any other questions/comments/concerns/thoughts?

risdenk commented:
Merged in https://github.com/apache/solr/commit/cb2e58f6569704c45387a8d094625ba8e014776b

risdenk commented:
Big shout out to @warperwolf who did 99% of the heavy lifting here.
